tests/federation_room_join_partial_state_test.go

```diff
@@ -613,26 +623,28 @@ func testReceiveEventDuringPartialStateJoin(
 		)
 		var respStateIDs gomatrixserverlib.RespStateIDs
 		if err := psjResult.Server.SendFederationRequest(deployment, stateReq, &respStateIDs); err != nil {
-			httpErr, ok := err.(gomatrix.HTTPError)
-			t.Logf("%v", httpErr)
-			if ok && httpErr.Code == 403 {
-				stateIdsResponseChan <- nil
-				return
-			}
-			t.Errorf("/state_ids request returned non-200: %s", err)
-			return
+			stateIdsResultChan <- StateIDsResult{Error: err}
+		} else {
+			stateIdsResultChan <- StateIDsResult{RespStateIDs: respStateIDs}
 		}
-		stateIdsResponseChan <- &respStateIDs
 	}()
 
 	select {
 	case <-time.After(1 * time.Second):
 		t.Logf("/state_ids request for event %s blocked as expected", event.EventID())
-		defer func() { <-stateIdsResponseChan }()
+		// read from the channel and discard the result, so that the goroutine above doesn't block
+		// indefinitely on the send
```

Comment:
(I'm a bit surprised that closing the channel isn't sufficient for this, but fair enough)

Reply:
We get a `panic: send on closed channel` without this. To clarify:

no close and no read -> complement times out after 10 minutes
defer close + no read -> panic
defer close + defer read -> complement is happy

Reply:
I don't really see why this is done in a goroutine at all - is this because `SendFederationRequest` can block indefinitely? There is a default [30s timeout](https://github.com/matrix-org/gomatrixserverlib/blob/fc2ea1e62e4606cfc415fb0da826ce69586a51d6/client.go#L39) - if that is too long then you need to send a custom context with a timeout here - https://github.com/matrix-org/complement/blob/d842670c55a2cdc61ec12e1640bc593b3f982542/internal/federation/server.go#L227 - currently this is `context.Background()` so it won't ever expire, but you can change that easily by doing:
```go
ctx, cancel := context.WithTimeout(context.Background(), time.Second)
defer cancel()
httpClient.DoRequestAndParseResponse(ctx, httpReq, resBody)
```

This will require a change to Complement's API to:
```go
func (s *Server) SendFederationRequest(ctx context.Context, deployment *docker.Deployment, req gomatrixserverlib.FederationRequest, resBody interface{}) error {
```

but that's fine.

Reply:
Thanks for the suggestion, that's a lot cleaner!